source/engines/rollEngine.js

```diff
@@ -1,548 +1,771 @@
+		} else if (/^\d+d(\%|\d+)$/.test(parsingString)) { //parse multiple die roll
+			var dPos = parsingString.search('d');
+			return new MultiDieResultSet(parsingString.slice(0,dPos),parsingString.slice(dPos - parsingString.length));
+		} else if (/^-?\d+$/.test(parsingString)) { //parse number result
+			return new SingleResultSet(parsingString,false);
+		} else if (/^1?u?l\[.*\]/.test(parsingString)) { //parse single list roll
+			var bracketList = parsingString.replace(/^1?u?l/,'');
+			return new ListResultSet(1, this.#parseStringList(bracketList));
+		} else if (/^\d+l\[.*\]/.test(parsingString)) { // parse multiple list pick
+			var bracketList = parsingString.replace(/^\d+l/,'');
+			var num = parsingString.replace(/l\[.*\]/,'');
+			return new ListResultSet(num, this.#parseStringList(bracketList));
+		} else if (/^\d+ul\[.*\]/.test(parsingString)) { // parse multiple list pick with unique selections
+			var bracketList = parsingString.replace(/^\d+ul/,'');
+			var num = parsingString.replace(/ul\[.*\]/,'');
+			var forceUnique = !(num === '' || Number.parseInt(num) === 1);
+			return new ListResultSet(num, this.#parseStringList(bracketList), forceUnique);
+		}
+		/*
+		 * Find the appropriate symbol, right to left
+		 * If the symbol is not surrounded by parens, turn into a TwoOpResultSet
+		 * Otherwise, continue try again
+		 * If the conditions are not met, try the next symbol set
+		 */
+		//Handle addition/subtraction
+		var addSubPattern = /[\+\-]/;
+		var multPattern = /\*/;
+		var hasListPattern = /\d*u?l\[.*\]/;
+		if(!multPattern.test(parsingString) && !hasListPattern.test(parsingString) && addSubPattern.test(parsingString)) {
+			var toPureSums = parsingString.replace(/\-/g,'+-');
+			var parsableList = toPureSums.split('+');
+			return new SumSeriesResultSet(parsableList);
+		}
+		if (addSubPattern.test(parsingString)) {
+			for (var i = parsingString.length - 1; i >= 0; i-=1) {
+				if (addSubPattern.test(parsingString[i]) && !this.#surroundedByParens(parsingString,i)) {
+					return new TwoOpResultSet(parsingString.slice(0,i),parsingString.slice(i+1,parsingString.length),parsingString[i]);
+				}
+			}
+		}
+		if (multPattern.test(parsingString)) {
+			for (var i = parsingString.length - 1; i >= 0; i-=1) {
+				if (multPattern.test(parsingString[i]) && !this.#surroundedByParens(parsingString,i)) {
+					return new TwoOpResultSet(parsingString.slice(0,i),parsingString.slice(i+1,parsingString.length),parsingString[i]);
+				}
+			}
+		}
+		//string does not match any expected combination - throw error
+		throw `Part of the input to parse into a roll is malformed: ${parsingString}`;
+	}
+
+	getResults() { throw `ResultSet function 'getResults' not properly implemented by child class ${this.constructor.name}.`; } // Implemented by child classes

```

Comment:
I am stealing the "throw `function not properly implemented by child class`" for use elsewhere.

Reply:
It is definitely a convenient way of enforcing `interface`s